snapcraft/remote/utils.py

```diff
@@ -76,3 +78,27 @@ def _compute_hash(directory: Path) -> str:
 
     all_hashes = "".join(hashes).encode()
     return md5(all_hashes).hexdigest()  # noqa: S324 (insecure-hash-function)
+
+
+def rmtree(directory: Path) -> None:
+    """Cross-platform rmtree implementation.
+
+    :param directory: Directory to remove.
+    """
+    shutil.rmtree(
+        str(directory.resolve()),
+        onerror=_remove_readonly,  # type: ignore
+    )
+
+
+def _remove_readonly(func, filepath):
```

Comment:
I can't figure this one out (I see why it was not tested in snapcraft_legacy).

I feel like I'm close with:
```python
def test_rmtree_readonly(mocker):
    """Remove a directory tree that contains a read-only file."""
    Path("root-dir").mkdir()
    Path("root-dir/test").touch()

    mocker.patch("os.unlink", side_effect=[OSError])

    rmtree(Path("root-dir"))

    assert not Path("root-dir").exists()
```

But it's not working and I don't like how dependent the test is on shutil's implementation.

Reply:
Yeah I don't have any good ideas about it, but the update is an improvement. Since this was brought from legacy I think it's not a stopper until we actually get this in the real world, and frankly if you'd excluded this function altogether I probably wouldn't have even thought about it.